# Lab Instructions

Find a dataset that interests you. I'd recommend starting on [Kaggle](https://www.kaggle.com/). Read through all of the material about the dataset and download a .CSV file.

1. Write a short summary of the data.  Where did it come from?  How was it collected?  What are the features in the data?  Why is this dataset interesting to you?  

2. Identify 5 interesting questions about your data that you can answer using Pandas methods.  

3. Answer those questions!  You may use any method you want (including LLMs) to help you write your code; however, you should use Pandas to find the answers.  LLMs will not always write code in this way without specific instruction.  

4. Write the answer to your question in a text box underneath the code you used to calculate the answer.



The data that I'm using is the race data from formula 1 for the years 1950-2026. This data has been gathered and posted to the public domain by Chris Newell. It contains basically every stat regarding formula1 but I'm choosing just to cover race results and I added the drivers so that I can see who the drivers are and what races they were at that top the stats I'll be looking at. This dataset is interesting to me as I enjoy watching F1 and knowing more about something I enjoy can't ever be a bad thing.

1. Which drivers have driven the most laps in their career?
2. Which drivers have had the best average finishing position throughout their careers with atleast 50 race finishes? 
3. Which drivers have had the most pole positions (grid as 1)?
4. which constructors have had the most unique drivers race for them?
5. How many races have been won with a gap of less than 5 seconds?

In [29]:
import pandas as pd
results_df     = pd.read_csv("results.csv")

#replace all \n null markers with NaN to make sure calculations work later
results_df.replace('\\N', pd.NA, inplace=True)
results_df['laps']         = pd.to_numeric(results_df['laps'],         errors='coerce')
results_df['position']     = pd.to_numeric(results_df['position'],     errors='coerce')
results_df['grid']         = pd.to_numeric(results_df['grid'],         errors='coerce')
results_df['milliseconds'] = pd.to_numeric(results_df['milliseconds'], errors='coerce')
results_df.head()

,resultId,raceId,driverId,constructorId,number,grid,position,positionText,positionOrder,points,laps,time,milliseconds,fastestLap,rank,fastestLapTime,fastestLapSpeed,statusId
0,1,18,1,1,22,1.0,1.0,1,1,10.0,58,1:34:50.616,5690616.0,39,2,1:27.452,218.3,1
1,2,18,2,2,3,5.0,2.0,2,2,8.0,58,+5.478,5696094.0,41,3,1:27.739,217.586,1
2,3,18,3,3,7,7.0,3.0,3,3,6.0,58,+8.163,5698779.0,41,5,1:28.090,216.719,1
3,4,18,4,4,5,11.0,4.0,4,4,5.0,58,+17.181,5707797.0,58,7,1:28.603,215.464,1
4,5,18,5,1,23,3.0,5.0,5,5,4.0,58,+18.014,5708630.0,43,1,1:27.418,218.385,1


In [11]:
constructors_df = pd.read_csv('constructors.csv')
constructors_df.head()

,constructorId,constructorRef,name,nationality,url
0,1,mclaren,McLaren,British,http://en.wikipedia.org/wiki/McLaren
1,2,bmw_sauber,BMW Sauber,German,http://en.wikipedia.org/wiki/BMW_Sauber
2,3,williams,Williams,British,http://en.wikipedia.org/wiki/Williams_Grand_Pr...
3,4,renault,Renault,French,http://en.wikipedia.org/wiki/Renault_in_Formul...
4,5,toro_rosso,Toro Rosso,Italian,http://en.wikipedia.org/wiki/Scuderia_Toro_Rosso


In [6]:
drivers_df = pd.read_csv('drivers.csv')
drivers_df.head()

,driverId,driverRef,number,code,forename,surname,dob,nationality,url
0,1,hamilton,44,HAM,Lewis,Hamilton,1985-01-07,British,http://en.wikipedia.org/wiki/Lewis_Hamilton
1,2,heidfeld,\N,HEI,Nick,Heidfeld,1977-05-10,German,http://en.wikipedia.org/wiki/Nick_Heidfeld
2,3,rosberg,6,ROS,Nico,Rosberg,1985-06-27,German,http://en.wikipedia.org/wiki/Nico_Rosberg
3,4,alonso,14,ALO,Fernando,Alonso,1981-07-29,Spanish,http://en.wikipedia.org/wiki/Fernando_Alonso
4,5,kovalainen,\N,KOV,Heikki,Kovalainen,1981-10-19,Finnish,http://en.wikipedia.org/wiki/Heikki_Kovalainen


In [12]:
races_df = pd.read_csv('races.csv')
races_df.head()

,raceId,year,round,circuitId,name,date,time,url,fp1_date,fp1_time,fp2_date,fp2_time,fp3_date,fp3_time,quali_date,quali_time,sprint_date,sprint_time
0,1,2009,1,1,Australian Grand Prix,2009-03-29,06:00:00,http://en.wikipedia.org/wiki/2009_Australian_G...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
1,2,2009,2,2,Malaysian Grand Prix,2009-04-05,09:00:00,http://en.wikipedia.org/wiki/2009_Malaysian_Gr...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
2,3,2009,3,17,Chinese Grand Prix,2009-04-19,07:00:00,http://en.wikipedia.org/wiki/2009_Chinese_Gran...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
3,4,2009,4,3,Bahrain Grand Prix,2009-04-26,12:00:00,http://en.wikipedia.org/wiki/2009_Bahrain_Gran...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
4,5,2009,5,4,Spanish Grand Prix,2009-05-10,12:00:00,http://en.wikipedia.org/wiki/2009_Spanish_Gran...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N


In [23]:
mostlaps_df = results_df.groupby('driverId')['laps'].sum().reset_index()

mostlaps_df = mostlaps_df.merge(drivers_df[['driverId', 'forename', 'surname']], on='driverId')
mostlaps_df['name'] = mostlaps_df['forename'] + ' ' + mostlaps_df['surname']

mostlaps_df = mostlaps_df[['name', 'laps']].sort_values('laps', ascending=False)
top10_ids = mostlaps_df.head(10)['driverId']
mostlaps_df.head(10)

,name,laps
3,Fernando Alonso,23189
0,Lewis Hamilton,21765
7,Kimi Räikkönen,18618
29,Michael Schumacher,16824
21,Rubens Barrichello,16642
19,Sebastian Vettel,16426
17,Jenson Button,16274
813,Sergio Pérez,15654
12,Felipe Massa,14852
815,Daniel Ricciardo,14129


1.Which drivers have driven the most laps in their career?
The drivers who have driven the most laps in their career are Fernando Alonso with 23189 laps and Lewis Hamilton with 21765 laps.
I was curious how long this was in terms of time so I'm gonna also answer that question below this one.

In [26]:
results_df['milliseconds'] = pd.to_numeric(results_df['milliseconds'], errors='coerce')
top10_time = results_df[results_df['driverId'].isin(top10_ids)].groupby('driverId')['milliseconds'].sum().reset_index()
top10_time = top10_time.merge(drivers_df[['driverId', 'forename', 'surname']], on='driverId')
top10_time['name'] = top10_time['forename'] + ' ' + top10_time['surname']
top10_time['total_hours'] = (top10_time['milliseconds'] / 3600000).round(2)

print(results_df[results_df['driverId'].isin(top10_ids)]['milliseconds'].isnull().groupby(results_df['driverId']).sum())
top10_time[['name', 'total_hours']].sort_values('total_hours', ascending=False)

driverId
1       46
4      152
8      130
13     104
18     148
20      73
22     181
30      89
815    114
817    106
Name: milliseconds, dtype: int64


,name,total_hours
0,Lewis Hamilton,541.63
1,Fernando Alonso,448.58
5,Sebastian Vettel,370.40
2,Kimi Räikkönen,348.37
7,Michael Schumacher,347.10
8,Sergio Pérez,276.90
3,Felipe Massa,269.01
4,Jenson Button,261.27
9,Daniel Ricciardo,247.51
6,Rubens Barrichello,231.13


1.5. How long have the drivers with the most laps spent on track?
It would seem the data is not super complete for this type of assessment but it's still somewhat interesting to look at regardless and Lewis having no missing data at 541 hours spent during races seems like a really long time to be under the load that a F1 car puts on a persons body.

In [31]:
avgfin_df = results_df.dropna(subset=['position'])
avgfin_df = avgfin_df.groupby('driverId').agg(avg_pos=('position', 'mean'), races=('position', 'count')).reset_index()
avgfin_df = avgfin_df[avgfin_df['races'] >= 50]
avgfin_df = avgfin_df.merge(drivers_df[['driverId', 'forename', 'surname']], on='driverId')
avgfin_df['name'] = avgfin_df['forename'] + ' ' + avgfin_df['surname']
avgfin_df['avg_pos'] = avgfin_df['avg_pos'].round(2)
avgfin_df = avgfin_df[['name', 'avg_pos', 'races']].sort_values('avg_pos')
avgfin_df.head(10)

,name,avg_pos,races
69,Jackie Stewart,2.95,63
39,Alain Prost,2.96,143
36,Ayrton Senna,3.15,108
85,Max Verstappen,3.66,206
20,Michael Schumacher,3.70,241
75,Jim Clark,3.84,50
35,Nigel Mansell,3.92,98
0,Lewis Hamilton,3.98,353
43,Nelson Piquet,4.14,122
21,Juan Pablo Montoya,4.21,62


2.Which drivers have had the best average finishing position throughout their careers with atleast 50 race finishes? 
The leaders in this list are almost all new names from the most laps list with the exception of lewis hamilton which helps explain why most people consider him the greatest F1 driver of all time. This also supports the conclusion that you don't have to be dominant to have a long career in F1.

In [32]:
mostpoles_df = results_df[results_df['grid'] == 1].groupby('driverId').size().reset_index(name='poles')
mostpoles_df = mostpoles_df.merge(drivers_df[['driverId', 'forename', 'surname']], on='driverId')
mostpoles_df['name'] = mostpoles_df['forename'] + ' ' + mostpoles_df['surname']
mostpoles_df = mostpoles_df[['name', 'poles']].sort_values('poles', ascending=False)
mostpoles_df.head(10)


,name,poles
0,Lewis Hamilton,104
16,Michael Schumacher,68
25,Ayrton Senna,65
12,Sebastian Vettel,57
104,Max Verstappen,48
69,Jim Clark,34
28,Alain Prost,33
24,Nigel Mansell,32
2,Nico Rosberg,30
84,Juan Fangio,29


3.Which drivers have had the most pole positions (grid as 1)?
Once again we see Lewis hamilton in the top 10. This list has alot of similarities with the average finishing position list which makes alot of sense as starting further up the grid would naturally lead to a better finishing position.

In [34]:
mostdrivers_df = results_df.groupby('constructorId')['driverId'].nunique().reset_index(name='unique_drivers')
mostdrivers_df = mostdrivers_df.merge(constructors_df[['constructorId', 'name']], on='constructorId')
mostdrivers_df = mostdrivers_df[['name', 'unique_drivers']].sort_values('unique_drivers', ascending=False)
mostdrivers_df.head(10)

,name,unique_drivers
5,Ferrari,99
102,Maserati,92
110,Kurtis Kraft,87
64,BRM,70
2,Williams,63
31,Team Lotus,61
166,Cooper-Climax,58
0,McLaren,55
168,Lotus-Climax,55
33,Brabham,49


4.which constructors have had the most unique drivers race for them?
This one doesn't really show any sort of meaningful stat related to the performance of the teams but I think it's somewhat interesting to look at. It's cool to see Ferrari, Williams, and Mclaren on the list as well as these are the only 3 teams on this list that are still currently on the grid.

In [37]:
winners_df    = results_df[results_df['positionOrder'] == 1][['raceId', 'milliseconds']].rename(columns={'milliseconds': 'winner_ms'})
runnerups_df  = results_df[results_df['positionOrder'] == 2][['raceId', 'milliseconds']].rename(columns={'milliseconds': 'p2_ms'})
closewins_df  = winners_df.merge(runnerups_df, on='raceId').dropna(subset=['winner_ms', 'p2_ms'])
closewins_df['gap_s'] = (closewins_df['p2_ms'] - closewins_df['winner_ms']) / 1000
closewins_df  = closewins_df[closewins_df['gap_s'] < 5]
closewins_df  = closewins_df.merge(races_df[['raceId', 'year', 'name']], on='raceId')
closewins_df['gap_s'] = closewins_df['gap_s'].round(3)
closewins_df  = closewins_df[['year', 'name', 'gap_s']].sort_values('gap_s')
print(closewins_df.count())
closewins_df.head(10)

year     374
name     374
gap_s    374
dtype: int64


,year,name,gap_s
190,1971,Italian Grand Prix,0.010
54,2002,United States Grand Prix,0.011
132,1986,Spanish Grand Prix,0.014
141,1982,Austrian Grand Prix,0.050
194,1969,Italian Grand Prix,0.080
210,1961,French Grand Prix,0.100
220,1954,French Grand Prix,0.100
65,2000,Canadian Grand Prix,0.174
345,2023,Australian Grand Prix,0.179
46,2002,Austrian Grand Prix,0.182


5.How many races have been won with a gap of less than 5 seconds?
It would appear that the answer to this question is 374 which is alot more than I was expecting but I guess 5 seconds is a pretty big amount of time. The closest races are really interesting as well with 3 of them being decided by just one hundredth of a second which is crazy close.